src/ports/postgres/modules/sample/balance_sample.py_in

```diff
@@ -468,81 +544,107 @@ def balance_sample(schema_madlib, source_table, output_table, class_col,
-        target_class_sizes = _get_target_level_counts(sampling_strategy_str,
-                                                      parsed_class_sizes,
-                                                      actual_level_counts,
-                                                      output_table_size)
-
-        undersample_level_dict, oversample_level_dict, nosample_level_dict = \
-            _get_sampling_strategy_specific_dict(target_class_sizes)
-
-        # Get subqueries for each sampling strategy, so that they can be used
-        # together in one big query.
-
-        # Subquery that will be used to get rows as is for those class levels
-        # that need no sampling.
-        nosample_subquery = _get_nosample_subquery(
-            new_source_table, class_col, nosample_level_dict.keys())
-        # Subquery that will be used to sample those class levels that
-        # have to be oversampled.
-        oversample_subquery = _get_with_replacement_subquery(
-            schema_madlib, new_source_table, source_table_columns, class_col,
-            actual_level_counts, oversample_level_dict)
-        # Subquery that will be used to sample those class levels that
-        # have to be undersampled. Undersampling supports both with and without
-        # replacement, so fetch the appropriate subquery.
-        if with_replacement:
-            undersample_subquery = _get_with_replacement_subquery(
-                schema_madlib, new_source_table, source_table_columns, class_col,
-                actual_level_counts, undersample_level_dict)
-        else:
-            undersample_subquery = _get_without_replacement_subquery(
-                schema_madlib, new_source_table, source_table_columns, class_col,
-                actual_level_counts, undersample_level_dict)
-
-        # Merge the three subqueries using a UNION ALL clause.
-        union_all_subquery = ' UNION ALL '.join(
-            ['({0})'.format(subquery)
-             for subquery in [undersample_subquery, oversample_subquery, nosample_subquery]
-             if subquery])
-
-        final_query = """
-                CREATE TABLE {output_table} AS
-                SELECT row_number() OVER() AS {new_col_name}, *
+        grp_col_str, grp_cols = get_grouping_col_str(
+            schema_madlib, 'Balance sample', [NEW_ID_COLUMN, class_col],
+            source_table, grouping_cols)
+        actual_grp_level_counts = _get_level_frequency_distribution(
+            new_source_table, class_col, grp_col_str)
+
+        is_output_created = False
+        n_grp_values = len(actual_grp_level_counts)
+        grp_cols_list = grp_col_str.split(',') if grp_cols else []
+
+        # for each group
+        for grp_vals, actual_level_counts in actual_grp_level_counts.items():
+            target_class_sizes = _get_target_level_counts(
+                sampling_strategy_str, parsed_class_sizes,
+                actual_level_counts, output_table_size, n_grp_values)
+
```

Comment:
Following the above comment, we can then have a new for loop that starts here, which will be executed only if there is no error condition arising due to `sampling_strategy_str`.